### PR TITLE
Fix various memory leaks

### PR DESCRIPTION
**Pull request by JasonMoho**
This change fixes memory leaks present in the C++ code base. 

Changes:
- Fix major memory leaks in AsyncWriteBlock and LookaheadBlock causing training to run out of memory on large scale dataset while using the PartitionBuffer with prefetching on.
- Destructors added for multiple classes
- Fixes various small memory leaks
- Added GitHub Actions workflow which builds Marius with ASAN and tests for memory leaks in three cases
    - Host memory storage for all program data
    - Partition Buffer storage for embeddings 
    - FlatFile storage for edges
- Additional test data and configuration files under test/test_data and test/test_configs for the workflow

After these changes ASAN finds no memory leaks.

### CODE REVIEW

File: src/cpp/include/io.h [hunk 1]
```diff
@@ -37,4 +37,6 @@ tuple<Storage *, Storage *, Storage *, Storage *, Storage *, Storage *, Storage
 
 tuple<Storage *, Storage *, Storage *, Storage *> initializeEval();
 
+
```
Comment (shivaram) [inline src/cpp/include/io.h:40]: Just blank lines. revert?

File: src/cpp/src/marius.cpp [hunk 2]
```diff
@@ -96,6 +96,99 @@ void marius(int argc, char *argv[]) {
...
+
+        // garbage collect
+        delete trainer;
+        delete train_set;
+        if (will_evaluate) {
+            delete evaluator;
+            delete eval_set;
+        }
+
+        switch (marius_options.storage.edges) {
```
Comment (shivaram) [inline src/cpp/src/marius.cpp:109]: this code block seems to be repeated three times -- maybe a good candidate to put inside a util function or something like that?

**Review by shivaram — APPROVED** [review]
Nice work!

File: src/cpp/src/marius.cpp (hunk 2 above)
Comment (JasonMoho) [reply src/cpp/src/marius.cpp]: The code blocks are slightly different, gotta handle the storage for the edges, node embedding parameters, and relation embedding parameters a little differently. I tossed these into a util function to make it cleaner.

**Pull request merged**